libs/thunk/windows/platform.cpp

```diff
@@ -131,6 +131,13 @@ _close(int file_descriptor)
     return ::_close(file_descriptor);
 }
 
+int
+_dup(int file_descriptor)
+{
+    _invalid_parameter_suppression suppress;
+    return ::_dup(file_descriptor);
```

Comment:
nit: does `_dup` also duplicate the underlying handle? If not, will this cause an issue when both the fds are eventually closed?

Reply:
Good question, https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/dup-dup2?view=msvc-170 doesn't specifically say. The test I added in https://github.com/microsoft/ebpf-for-windows/pull/4116/files#diff-c336aef1a5b6b610ede7ec16ac04f798e3620811eb0416754a40f2c741e4ece2R569 indicates that things work as expected.

I suspect that the runtime does some refcounting behind the scenes, since it can't duplicate the handle.

Reply:
I think `dup` does not duplicate the underlying handle. Closing an fd closes the underlying handle. If you try to call close on a dup fd, that fails as the handle is already closed. It asserts in chk bits while it likely fails silently in return.

Consider
`_get_handle_from_file_descriptor()` followed by `Platform::DuplicateHandle()` and then generating the fd of the duplicated handle.

Alternatively, just not attempt to close a dupe fd.

Reply:
Thanks for helping me debug this! Digging a bit deeper, this is the parameter validation that fails (in `_close_internal`):

```
_UCRT_VALIDATE_CLEAR_OSSERR_RETURN(ptd, (_osfile(fh) & FOPEN), EBADF, -1);
```

Clearly, this is because I've closed that same fd before (and kind of what I expected to happen). If the problem is really that the handle is closed, wouldn't the error be generated further down the stack in CloseHandle or similar?

I'm also puzzled why `_invalid_parameter_suppression` isn't kicking in? (I think there is a race condition here, but not sure how that would be triggered in the test.)

Reply:
I changed the error reporting a little, and tried to reproduce locally:

```
minkernel\crts\ucrt\src\appcrt\lowio\close.cpp(56) : Assertion failed: (_osfile(fh) & FOPEN)
C:\Users\lorenz\ebpf-for-windows\tests\api_test\api_test.cpp(569): PASSED:
  REQUIRE( ebpf_close_fd(map_fd2) == EBPF_FAILED )
with expansion:
  17 == 17

minkernel\crts\ucrt\src\appcrt\lowio\osfinfo.cpp(261) : Assertion failed: _osfile(fh) & FOPEN
C:\Users\lorenz\ebpf-for-windows\tests\api_test\api_test.cpp(570): PASSED:
  REQUIRE( bpf_map_lookup_elem(map_fd2, &key, &value) == -9 )
with expansion:
  -9 == -9
```

-> We can see that the assertion in ucrt fails, but at the same time the unit test is successful because we really do return EBADF.

Reply:
The problem is _wer_report::terminate_hook. I'll work around this for now. Overall I'm not sure that terminating the process is the best action, because it completely swallows all error output.